mlir/include/mlir/Dialect/MPI/IR/MPIOps.td

```diff
@@ -0,0 +1,129 @@
+//===- MPI.td - Message Passing Interface Ops ---------*- tablegen -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MPI_OPS
+#define MPI_OPS
+
+include "mlir/Dialect/MPI/IR/MPIBase.td"
+
+class MPI_Op<string mnemonic, list<Trait> traits = []> :
+    Op<MPI_Dialect, mnemonic, traits>;
+
+//===----------------------------------------------------------------------===//
+// InitOp
+//===----------------------------------------------------------------------===//
+
+def MPI_InitOp : MPI_Op<"init", [
+    
+  ]> {
+  let summary = "Initialize the MPI library, equivalent to `MPI_Init(NULL, NULL)`";
+  let description = [{
+    This operation must preceed most MPI calls (except for very few exceptions,
+    please consult with the MPI specification on these).
+
+    Passing &argc, &argv is not supported currently.
+    Inspecting the functions return value (error code) is also not supported.
+  }];
+
+  let assemblyFormat = "attr-dict";
+}
+
+//===----------------------------------------------------------------------===//
+// CommRankOp
+//===----------------------------------------------------------------------===//
+
+def MPI_CommRankOp : MPI_Op<"comm_rank", [
+    
+  ]> {
+  let summary = "Get the current rank, equivalent to `MPI_Comm_rank(MPI_COMM_WORLD, &rank)`";
+  let description = [{
+    Communicators other than `MPI_COMM_WORLD` are not supprted for now.
+    Inspecting the functions return value (error code) is also not supported.
+  }];
+
+  let results = (outs
+    I32:$result
+  );
+
+  let assemblyFormat = "attr-dict `:` type($result)";
+}
+
+
+//===----------------------------------------------------------------------===//
+// SendOp
+//===----------------------------------------------------------------------===//
+
+def MPI_SendOp : MPI_Op<"send", [
+    
+  ]> {
+  let summary = "Equivalent to `MPI_Send(ptr, size, dtype, dest, tag, MPI_COMM_WORLD)`";
+  let description = [{
+    MPI_Send performs a blocking send of `size` elements of type `dtype` to rank `dest`.
+    The `tag` value and communicator enables the library to determine the matching of
+    multiple sends and receives between the same ranks.
```

Comment:
:+1: 
sounds perfect

Reply:
Does this always have to be used with MPI_Recv? If so, maybe mention it here. I guess even if it's jsut "usually" or "mostly" the case, it might be worth cross referencing.

Reply:
I think that one should defer to the MPI standard for this kind of thing. What you want to know in this dialect is how the dialect operations map to the MPI calls.

FWIW, `MPI_Send` does not always have to be matched with `MPI_Recv` (it can also be matched with `MPI_Irecv`, or, if the `dest = MPI_PROCESS_NULL` then it doesn't have to be matched with anything).

Reply:
Maybe we can have a high-level summary inlined in the documentation and somewhere add a reference to the MPIR standard as @wence- suggested.